lib/Transforms/ConvertSecretExtractToStaticExtract/ConvertSecretExtractToStaticExtract.cpp

```diff
@@ -0,0 +1,212 @@
+#include "mlir/include/mlir/IR/Operation.h"              // from @llvm-project
+#include "mlir/include/mlir/IR/PatternMatch.h"           // from @llvm-project
+#include "mlir/include/mlir/IR/Value.h"                  // from @llvm-project
+#include "mlir/include/mlir/Interfaces/SideEffectInterfaces.h"  // from @llvm-project
+#include "mlir/include/mlir/Support/LLVM.h"           // from @llvm-project
+#include "mlir/include/mlir/Support/LogicalResult.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
+
+namespace mlir {
+namespace heir {
+
+#define GEN_PASS_DEF_CONVERTSECRETEXTRACTTOSTATICEXTRACT
+#include "lib/Transforms/ConvertSecretExtractToStaticExtract/ConvertSecretExtractToStaticExtract.h.inc"
+
+struct SecretExtractToStaticExtractConversion
+    : OpRewritePattern<tensor::ExtractOp> {
+  using OpRewritePattern<tensor::ExtractOp>::OpRewritePattern;
+
+ public:
+  SecretExtractToStaticExtractConversion(DataFlowSolver *solver,
+                                         MLIRContext *context)
+      : OpRewritePattern(context), solver(solver) {}
+
+  LogicalResult matchAndRewrite(tensor::ExtractOp extractOp,
+                                PatternRewriter &rewriter) const override {
+    // TODO(#866): Add support for reads on multi-dimensional tensors
+    if (extractOp.getTensor().getType().getRank() != 1) {
+      extractOp->emitWarning()
+          << "Currently, transformation only supports 1D tensors:";
+      return failure();
+    }
+
+    auto index = extractOp.getIndices().front();
+
+    auto *indexSecretnessLattice =
+        solver->lookupState<SecretnessLattice>(index);
+
+    // If (1) lattice is not found or (2) lattice is set to unknown, apply
+    // transformation
+    if (!indexSecretnessLattice ||
+        !indexSecretnessLattice->getValue().isInitialized()) {
+      extractOp->emitWarning()
+          << "Secretness for tensor.extract index has not been set. "
+             "Conservatively, the transformation will be applied:";
+    }
+
+    auto indexSecretness = indexSecretnessLattice->getValue();
+
+    bool isIndexSecret =
+        !indexSecretness.isInitialized() ||
+        (indexSecretness.isInitialized() && indexSecretness.getSecretness());
+
+    // If index is not secret, no transformation is needed
+    if (!isIndexSecret) return failure();
+
+    ImplicitLocOpBuilder builder(extractOp->getLoc(), rewriter);
+
+    auto argType = extractOp.getResult().getType();
+
+    auto *extractOpLattice =
+        solver->lookupState<SecretnessLattice>(extractOp.getResult());
+
+    auto extractOpSecretness = extractOpLattice ? extractOpLattice->getValue() :
+                                                /*unknown*/ Secretness();
+
+    // Placeholder for extracted value
+    auto dummyValue = builder.create<arith::ConstantOp>(
+        argType, builder.getIntegerAttr(argType, 0));
```

Comment:
### While this is perfectly fine/correct, I think it might be nicer for things like vectorization to have the dummy value be `tensor.extract %t[c0]` (just makes everything more homogenous, especially if you consider what the unrolled loop would look like, and no need to create a dummy ptxt/ctxt value when lowering to actual schemes).

Reply:
That also means the secretness of `dummyValue` becomes `tensorSecretness`